core/src/main/java/org/opensearch/sql/executor/pagination/CanPaginateVisitor.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.executor.pagination;
+
+import org.opensearch.sql.ast.AbstractNodeVisitor;
+import org.opensearch.sql.ast.Node;
+import org.opensearch.sql.ast.expression.AllFields;
+import org.opensearch.sql.ast.tree.Project;
+import org.opensearch.sql.ast.tree.Relation;
+
+/**
+ * Use this unresolved plan visitor to check if a plan can be serialized by PaginatedPlanCache.
+ * If plan.accept(new CanPaginateVisitor(...)) returns true,
+ * then PaginatedPlanCache.convertToCursor will succeed. Otherwise, it will fail.
+ * The purpose of this visitor is to activate legacy engine fallback mechanism.
+ * Currently, the conditions are:
+ * - only projection of a relation is supported.
+ * - projection only has * (a.k.a. allFields).
+ * - Relation only scans one table
+ * - The table is an open search index.
```

Comment:
how to determine is an index? The core engine should have no knowledge about it.

Reply:
Right, we don't validate the index here. In case if table is not an OpenSearch index, the query would fail later.
For example, a failure on querying an OpenSearch system index:
```json
$ curl -XPOST http://localhost:9200/_plugins/_sql -H 'Content-Type: application/json' -d '{"query": "show tables like \"%\"", "fetch_size": 2 }'
{
  "error": {
    "reason": "There was internal problem at backend",
    "details": "'class org.opensearch.sql.opensearch.storage.system.OpenSearchSystemIndex' does not support pagination",
    "type": "UnsupportedOperationException"
  },
  "status": 503
}
```

Exception stack:
```
java.lang.UnsupportedOperationException: 'class org.opensearch.sql.opensearch.storage.system.OpenSearchSystemIndex' does not support pagination
        at org.opensearch.sql.storage.Table.createPagedScanBuilder(Table.java:98) ~[core-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.planner.optimizer.rule.CreatePagingTableScanBuilder.apply(CreatePagingTableScanBuilder.java:67) ~[core-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.planner.optimizer.rule.CreatePagingTableScanBuilder.apply(CreatePagingTableScanBuilder.java:23) ~[core-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.planner.optimizer.LogicalPlanOptimizer.internalOptimize(LogicalPlanOptimizer.java:84) ~[core-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.planner.optimizer.LogicalPlanOptimizer.optimize(LogicalPlanOptimizer.java:69) ~[core-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.planner.Planner.optimize(Planner.java:65) ~[core-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.planner.Planner.plan(Planner.java:41) ~[core-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.executor.QueryService.plan(QueryService.java:108) ~[core-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.executor.QueryService.lambda$executePlan$1(QueryService.java:75) ~[core-3.0.0.0-SNAPSHOT.jar:?]
        at java.util.Optional.ifPresentOrElse(Optional.java:203) ~[?:?]
        at org.opensearch.sql.executor.QueryService.executePlan(QueryService.java:72) [core-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.executor.QueryService.execute(QueryService.java:43) [core-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.executor.execution.PaginatedPlan.execute(PaginatedPlan.java:45) [core-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.opensearch.executor.OpenSearchQueryManager.lambda$submit$0(OpenSearchQueryManager.java:33) [opensearch-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.sql.opensearch.executor.OpenSearchQueryManager.lambda$withCurrentContext$1(OpenSearchQueryManager.java:47) [opensearch-3.0.0.0-SNAPSHOT.jar:?]
        at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:747) [opensearch-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
```